src/MongoDB.Driver.Core/Core/Authentication/ScramShaAuthenticator.cs

```diff
@@ -219,14 +228,27 @@ public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedF
                     Convert.FromBase64String(s),
                     int.Parse(i));
 
-                var clientKey = _hmac(encoding, saltedPassword, "Client Key");
+                byte[] clientKey;
+                byte[] serverKey;
+                var cacheKey = new CacheKey(saltedPassword, Convert.FromBase64String(s), int.Parse(i));
```

Comment:
I think the CacheKey should be based on the hashed password, not the salted password.

Furthermore, I think the computation of the salted password is the main thing we are trying to avoid (that's the computation that has a high iteration count). Therefore, the salted password computation should only occur when we have a cache miss. So lines 226-229 should be inside the else clause at lines 242-244.

Reply:
Should I be taking the password from _credential, then? Is the SecureString password hashed? I'll have to change the ScramCacheKey to take in a SecureString rather than a byte[] if I were to do that.
And I should probably put saltedPassword in the ScramCacheEntry then.

Reply:
Hm, `SecureString` offers its own set of challenges...

@jyemin Does the Java driver use the plaintext password as part of the key-tuple to the cache? Or is it using a hashed password? If the latter, is it using a built in `string.GetHashCode()` or is it using one of the SHA algorithms to hash the password before storing it as part of the key-tulpe in the cache?

Reply:
It uses a SHA hash of `password+salt`.  I think this is safe especially as salt is also an independent value in the cache key, and avoids storing the password un-hashed in the cache key.  You could include the SecureString in the cache key so long as there is a way to compare to SecureString instances for equality.

Reply:
Thanks @jyemin ! 
@ksadoff The hashing function should be available as `_h` of type `H`. To securely and temporarily extract the password from the SecureString, you should be able to follow this model:
https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Driver.Core/Core/Authentication/ScramSha256Authenticator.cs#L75-L101

Reply:
So to re-iterate, by following the above model, you should be able to extract the password from the credentials, concat the salt onto it, hash the result, and then use that as part of the CacheKey per @rstam 's suggestion.

Reply:
But I argue that there is NO need to concatenate the salt for run ANY hash function on the password.

Simply use the SecureString as part of the ScramCacheKey.